action/protocol/staking/protocol.go

```diff
@@ -412,16 +411,25 @@ func (p *Protocol) Commit(ctx context.Context, sm protocol.StateManager) error {
 
 // Handle handles a staking message
 func (p *Protocol) Handle(ctx context.Context, elp action.Envelope, sm protocol.StateManager) (*action.Receipt, error) {
-	featureWithHeightCtx := protocol.MustGetFeatureWithHeightCtx(ctx)
-	height, err := sm.Height()
+	csm, err := NewCandidateStateManager(sm)
 	if err != nil {
 		return nil, err
 	}
-	csm, err := NewCandidateStateManager(sm, featureWithHeightCtx.ReadStateFromDB(height))
+	view, err := sm.ReadView(_protocolID)
 	if err != nil {
 		return nil, err
 	}
-	return p.handle(ctx, elp, csm)
+	snapshot := view.(*ViewData).Snapshot()
+	receipt, err := p.handle(ctx, elp, csm)
+	if err != nil {
+		if err := view.(*ViewData).Revert(snapshot); err != nil {
+			return nil, errors.Wrap(err, "failed to revert view")
+		}
+		return receipt, err
+	}
+	sm.WriteView(_protocolID, csm.DirtyView())
```

Comment:
the `view` should already updated when handling the ation, thus we do not need to `WriteView` here